docs/design/2021-12-08-instance-scope.md

```diff
@@ -0,0 +1,224 @@
+
+```
++++ b/session/session.go
+@@ -1103,6 +1103,10 @@ func (s *session) GetGlobalSysVar(name string) (string, error) {
+                return "", variable.ErrUnknownSystemVar.GenWithStackByArgs(name)
+        }
+ 
++       if sv.HasInstanceScope() { // has INSTANCE scope only, not pure global
++               return "", errors.New("variable has only instance scope and no GetGlobal func. Not sure how to handle yet.")
++       }
++
+        sysVar, err := domain.GetDomain(s).GetGlobalVar(name)
+        if err != nil {
+                // The sysvar exists, but there is no cache entry yet.
+@@ -1121,6 +1125,7 @@ func (s *session) GetGlobalSysVar(name string) (string, error) {
+ }
+ 
+ // SetGlobalSysVar implements GlobalVarAccessor.SetGlobalSysVar interface.
++// it is used for setting instance scope as well.
+ func (s *session) SetGlobalSysVar(name, value string) (err error) {
+        sv := variable.GetSysVar(name)
+        if sv == nil {
+@@ -1132,6 +1137,9 @@ func (s *session) SetGlobalSysVar(name, value string) (err error) {
+        if err = sv.SetGlobalFromHook(s.sessionVars, value, false); err != nil {
+                return err
+        }
++       if sv.HasInstanceScope() { // skip for INSTANCE scope
++               return nil
++       }
+        if sv.GlobalConfigName != "" {
+                domain.GetDomain(s).NotifyGlobalConfigChange(sv.GlobalConfigName, variable.OnOffToTrueFalse(value))
+        }
+@@ -1148,6 +1156,9 @@ func (s *session) SetGlobalSysVarOnly(name, value string) (err error) {
+        if err = sv.SetGlobalFromHook(s.sessionVars, value, true); err != nil {
+                return err
+        }
++       if !sv.HasInstanceScope() { // skip for INSTANCE scope
++               return nil
++       }
+        return s.replaceGlobalVariablesTableValue(context.TODO(), sv.Name, value)
+ }
+```
+
+A "non-native" `INSTANCE` variable can be changed to a native one as follows:
+
+```
++       {Scope: ScopeInstance, Name: TiDBGeneralLog, Value: BoolToOnOff(DefTiDBGeneralLog), Type: TypeBool, skipInit: true, SetGlobal: func(s *SessionVars, val string) error {
+                ProcessGeneralLog.Store(TiDBOptOn(val))
+                return nil
+-       }, GetSession: func(s *SessionVars) (string, error) {
++       }, GetGlobal: func(s *SessionVars) (string, error) {
+                return BoolToOnOff(ProcessGeneralLog.Load()), nil
+        }},
+```
+
+This introduces a compatibility issue, since users who have previously configured instance scope with `SET [SESSION] tidb_general_log = x` will receive an error because the scope is wrong. This can be fixed by adding a legacy mode to the sysvar framework which allows `INSTANCE` scoped variables to be set with either `SET GLOBAL` or `SET SESSION`:
+
+```sql
+SET GLOBAL tidb_enable_legacy_instance_scope = 1;
+```
+
+We can default to `TRUE` for now, but should revisit this in the future.
+
+### Stage 2: Mapping All Configuration File Settings to System Variables
```

Comment:
Another point is, we have some mappings already: https://docs.pingcap.com/tidb/stable/dynamic-config#modify-tidb-configuration-online, for example, `log.enable-slow-log` vs `tidb_enable_slow_log`.

We need to think about how the existing mapping can be merged into this proposal

Reply:
The existing mappings can be retained, new mappings can be built based on a convention. The previous proposal for instance scope [has a mapping table ](https://docs.google.com/document/d/1RuajYAFVsjJCwCBpIYF--9jtgxDZZcz3cbP-iVuLxJI/edit?n=2020-09-15_Design_Doc_for_Instance_Variables#) - the rules for name conversion were not too complicated. It dates quickly though, as the system variables change.